dev/run-tests.py

```diff
@@ -0,0 +1,417 @@
+    test suite runner (hence the function name).
+    @return a set of unique test names"""
+    test_suite = list()
+
+    if amplab_jenkins:
+        run_cmd(['git', 'fetch', 'origin', 'master:master']).wait()
+
+        raw_output = subprocess.check_output(['git', 'diff', '--name-only', 'master'])
+        # remove any empty strings
+        changed_files = [f for f in raw_output.split('\n') if f]
+
+        # find any sql files
+        sql_files = [f for f in changed_files
+                     if multi_starts_with(f, 
+                                          "sql/", 
+                                          "bin/spark-sql", 
+                                          "sbin/start-thriftserver.sh")]
+
+        non_sql_files = set(changed_files).difference(set(sql_files))
+
+        if non_sql_files:
+            test_suite.append("CORE")
+        if sql_files:
+            print "[info] Detected changes in SQL. Will run Hive test suite."
+            test_suite.append("SQL")
+            if not non_sql_files:
+                print "[info] Detected no changes except in SQL. Will only run SQL tests."
+        return set(test_suite)
+    else:
+        # we aren't in the Amplab environment so merely run all tests
+        test_suite.append("CORE")
+        test_suite.append("SQL")
+        return set(test_suite)
+
+
+def set_title_and_block(title, err_block):
+    os.environ["CURRENT_BLOCK"] = error_codes[err_block]
+    line_str = "".join(['='] * 72)
+
+    print
+    print line_str
+    print title
+    print line_str
+
+
+def run_apache_rat_checks():
+    set_title_and_block("Running Apache RAT checks", "BLOCK_RAT")
+    run_cmd(["./dev/check-license"])
+
+
+def run_scala_style_checks():
+    set_title_and_block("Running Scala style checks", "BLOCK_SCALA_STYLE")
+    run_cmd(["./dev/lint-scala"])
+
+
+def run_python_style_checks():
+    set_title_and_block("Running Python style checks", "BLOCK_PYTHON_STYLE")
+    run_cmd(["./dev/lint-python"])
+
+
+def exec_maven(mvn_args=[]):
+    """Will call Maven in the current directory with the list of mvn_args passed
+    in and returns the subprocess for any further processing"""
+
+    return subprocess.Popen(["./build/mvn"] + mvn_args)
+
+
+def exec_sbt(sbt_args=[]):
+    """Will call SBT in the current directory with the list of mvn_args passed
+    in and returns the subprocess for any further processing"""
+
+    # NOTE: echo "q" is needed because sbt on encountering a build file
+    # with failure (either resolution or compilation) prompts the user for
+    # input either q, r, etc to quit or retry. This echo is there to make it
+    # not block.
+    echo_proc = subprocess.Popen(["echo", "\"q\n\""], stdout=subprocess.PIPE)
+    sbt_proc = subprocess.Popen(["./build/sbt"] + sbt_args,
+                                stdin=echo_proc.stdout,
+                                stdout=subprocess.PIPE)
+    echo_proc.wait()
```

Comment:
> I'm under the impression given various SO posts that communicate() will also buffer stdout and stderr which means we'd have no output on Jenkins for however it took for the build/sbt script to run.

This will only happen if you direct stdout and stderr to `subprocess.PIPE`. Otherwise they will go straight to the screen.

As for applying an output filter here (which I don't currently see for this section; is there supposed to be one?), the simplest way may be to use `pipes.quote()` and `shell=True` to safely quote the sbt arguments and just build a string that does what we want.

Something like:

```python
from pipes import quote

subprocess.Popen(
    """
    echo 'q\n' \
    | {sbt} \
    | grep -v ...
    """.format(sbt=' '.join([quote(a) for a in sbt_args])
)
```

You can also give it a shot by piping things between commands using `subprocess.PIPE`, but for involving piping like this, I've found it easier to just use `shell=True` and safe quoting.